packages/amplify_datastore_plugin_interface/lib/src/types/temporal/temporal_time.dart

```diff
@@ -0,0 +1,151 @@
+/// https://docs.aws.amazon.com/appsync/latest/devguide/scalars.html#appsync-defined-scalars
+class TemporalTime {
+  DateTime _dateTime;
+  int _nanoseconds; // DateTime only stores millisecond and microsecond
+  Duration _offset;
+  bool _hasZ =
+      false; // fromString only - Need to track between hh:mm:ss vs hh:mm:ssZ
+
+  /// Constructs a new TemporalTime at the current date.
+  static TemporalTime now() {
+    return TemporalTime(DateTime.now());
+  }
+
+  /// Constructs a new TemporalTime from a Dart DateTime
+  /// The date fields (year, month, day) are ignored
+  TemporalTime(DateTime dateTime) {
+    dateTime = dateTime.toUtc();
+    _dateTime = DateTime.utc(
+        1970, 1, 1, dateTime.hour, dateTime.minute, dateTime.second);
+    _nanoseconds =
+        (dateTime.millisecond * 1000000) + (dateTime.microsecond * 1000);
+  }
+
+  /// Constructs a new TemporalTime from a Dart DateTime and Duration
+  /// The date fields (year, month, day) are ignored
+  TemporalTime.withOffset(DateTime dateTime, Duration offset) {
+    dateTime = dateTime.toUtc();
+    _dateTime = DateTime.utc(
+        1970, 1, 1, dateTime.hour, dateTime.minute, dateTime.second);
+    _nanoseconds =
+        (dateTime.millisecond * 1000000) + (dateTime.microsecond * 1000);
+
+    if (offset.inDays > 0) {
+      throw new Exception("Cannot have an offset in days (hh:mm:ss)");
+    }
+
+    _offset = offset;
+  }
+
+  /// Constructs a new TemporalTime from a ISO8601 string adhering to the format:
+  ///  hh:mm
+  ///  hh:mmZ
+  ///  hh:mm:ss
+  ///  hh:mm:ssZ
+  ///  hh:mm:ss.sss
+  ///  hh:mm:ss.sssZ
+  ///  without Z:
+  ///    +hh:mm
+  ///    +hh:mm:ss
+  TemporalTime.fromString(String iso8601String) {
+    RegExp regExp = new RegExp(
+        r'^([0-2][0-9]):([0-5][0-9])(:([0-5][0-9])(.([0-9]{1,9}))?)?((z|Z)|((\+|-)[0-2][0-9]:[0-5][0-9](:[0-5][0-9])?))?',
+        caseSensitive: false,
+        multiLine: false);
+
+    // Validate
+    String regexString = regExp.stringMatch(iso8601String);
+    if (regexString != iso8601String) {
+      throw new Exception("invalid string input");
+    }
+
+    // Remove Z
+    regexString = iso8601String.replaceAll(new RegExp(r'(z|Z)'), '');
+
+    // Extract Time
+    var match = regExp.matchAsPrefix(regexString);
+
+    int hours = int.parse(match.group(1));
+    int minutes = int.parse(match.group(2));
+    int seconds = Temporal.getIntOr0(match.group(4));
+    _nanoseconds = Temporal.getIntOr0(match.group(6));
+    //int milliseconds =
+
+    _dateTime = DateTime.utc(1970, 1, 1, hours, minutes, seconds);
+
+    // Extract Offset
+    if (match.group(7) != null && match.group(7).isNotEmpty)
+      _offset = Temporal.stringToOffset(match.group(7));
+    else if (iso8601String.toLowerCase().contains("z")) _hasZ = true;
+  }
+
+  /// Return offset
+  Duration getOffset() {
+    return _offset;
+  }
+
+  /// Return DateTime with offset added
+  /// Date fields are set as January 1, 1970.
+  DateTime toDateTime() {
```

Comment:
Perhaps call this `getDateTime()` instead?  This method is not _converting_ the `TemporalDateTime`, it is simply returning the "instant" component of it.

Reply:
good call, changed the method name.

Reply:
Though could that confuse users if they thought they can call getDateTime and then modify the underlying TemporalDate/TemporalDatetime object as well?

Reply:
The way Dart DateTime works is if you call "add" or "subtract" a duration, a new DateTime object is returned instead of modifying the existing one.

Reply:
Do you think there should be a setDateTime method as well?

Reply:
> Though could that confuse users if they thought they can call getDateTime and then modify the underlying TemporalDate/TemporalDatetime object as well?

As a general rule, all objects that we return to customers, including these Temporal objects, should be immutable.    To achieve that, you'd actually want to return a copy of `_dateTime` in this method, rather than the _dateTime instance itself.

> The way Dart DateTime works is if you call "add" or "subtract" a duration, a new DateTime object is returned instead of modifying the existing one.

I do understand that - my point was that you shouldn't ever be adding/subtracting the duration from the instant (a.k.a. Dart DateTime), ever.  The instant, combined with the offset, can be used by the customer to compute hour, minute, second for display on a UI.  

> Do you think there should be a setDateTime method as well?

No, like mentioned above, all objects that we return to customers should be immutable.

Reply:
Sounds good, that works perfectly then.  The Dart DateTime class is inherently immutable as it has no methods that actually allow you to modify the underlying instance - thus I think no action is needed here then.  Returning the _dateTime instance variable is already returning an immutable.